src/ngraph/op/batch_norm.cpp

```diff
@@ -22,99 +22,163 @@
+    , m_epsilon(epsilon)
+{
+    constructor_validate_and_infer_types();
+}
+
+// DEPRECATED
+ngraph::op::BatchNormInference::BatchNormInference(double eps,
+                                                   std::shared_ptr<ngraph::Node> gamma,
+                                                   std::shared_ptr<ngraph::Node> beta,
+                                                   std::shared_ptr<ngraph::Node> input,
+                                                   std::shared_ptr<ngraph::Node> mean,
+                                                   std::shared_ptr<ngraph::Node> variance)
+    : Op("BatchNormInference", check_single_output_args({gamma, beta, input, mean, variance}))
+    , m_epsilon(eps)
+{
+    constructor_validate_and_infer_types();
+}
+
+void ngraph::op::BatchNormInference::validate_and_infer_types()
 {
     element::Type result_et;
     PartialShape result_batch_shape;
-    PartialShape result_channel_shape;
+    PartialShape result_channel_shape; // unused here
 
+    set_output_size(1);
     std::tie(result_et, result_batch_shape, result_channel_shape) =
         infer_batch_norm_forward(this,
                                  get_input_element_type(INPUT_DATA),
                                  get_input_element_type(INPUT_GAMMA),
                                  get_input_element_type(INPUT_BETA),
+                                 get_input_element_type(INPUT_MEAN),
+                                 get_input_element_type(INPUT_VARIANCE),
                                  get_input_partial_shape(INPUT_DATA),
                                  get_input_partial_shape(INPUT_GAMMA),
-                                 get_input_partial_shape(INPUT_BETA));
+                                 get_input_partial_shape(INPUT_BETA),
+                                 get_input_partial_shape(INPUT_MEAN),
+                                 get_input_partial_shape(INPUT_VARIANCE));
 
     set_output_type(0, result_et, result_batch_shape);
-    set_output_type(1, result_et, result_channel_shape);
-    set_output_type(2, result_et, result_channel_shape);
 }
 
 std::shared_ptr<ngraph::Node>
     ngraph::op::BatchNormInference::copy_with_new_args(const NodeVector& new_args) const
 {
     check_new_args_count(this, new_args);
     return std::make_shared<BatchNormInference>(
-        m_epsilon, new_args.at(0), new_args.at(1), new_args.at(2), new_args.at(3), new_args.at(4));
-}
-
-std::shared_ptr<ngraph::Node>
-    ngraph::op::BatchNormTraining::copy_with_new_args(const NodeVector& new_args) const
-{
-    check_new_args_count(this, new_args);
-    return std::make_shared<BatchNormTraining>(
-        m_epsilon, new_args.at(0), new_args.at(1), new_args.at(2));
+        new_args.at(2), new_args.at(0), new_args.at(1), new_args.at(3), new_args.at(4), m_epsilon);
 }
 
 ngraph::op::BatchNormTrainingBackprop::BatchNormTrainingBackprop(
-    double eps,
+    std::shared_ptr<ngraph::Node> input,
```

Comment:
I agree that it's better to have `input` first, but shouldn't we also reorder the `Input`s, i.e. change the order in which we pass things to `check_single_output_args`? "Too much work for this PR" is an acceptable answer.

Reply:
I'm seeing now that you addressed this in your initial comment. This works for me.